File: front/app/api/embeddingQueue.ts
import type { ApiResponse } from '~/types'
import { createQueueApi, type QueueApi } from './createQueueApi'

export interface EmbeddingQueueStatus {
  pending: number
  processing: number
  completed: number
  failed: number
  total: number
}

export interface EmbeddingQueueTask {
  id: number
  tag_id: number
  status: 'pending' | 'processing' | 'completed' | 'failed'
  error_message: string | null
  retry_count: number
  created_at: string
  started_at: string | null
  completed_at: string | null
  tag?: {
    id: number
    label: string
    category: string
    slug: string
  }
}

export interface EmbeddingQueueTasksResponse {
  tasks: EmbeddingQueueTask[]
  total: number
}

const queue: QueueApi<EmbeddingQueueTask> = createQueueApi<EmbeddingQueueTask>('embedding/queue')

export function useEmbeddingQueueApi() {
  return {
    getStatus(): Promise<ApiResponse<EmbeddingQueueStatus>> {
      return queue.getStatus() as Promise<ApiResponse<EmbeddingQueueStatus>>
    },

    getTasks(params?: { status?: string; limit?: number; offset?: number }): Promise<ApiResponse<EmbeddingQueueTasksResponse>> {
      return queue.getTasks(params) as Promise<ApiResponse<EmbeddingQueueTasksResponse>>
    },

    retryFailed(): Promise<ApiResponse<{ message: string }>> {
      return queue.retryFailed()
    },
  }
}

File: backend-go/internal/models/embedding_queue.go
package models

import "time"

const (
	EmbeddingQueueStatusPending    = "pending"
	EmbeddingQueueStatusProcessing = "processing"
	EmbeddingQueueStatusCompleted  = "completed"
	EmbeddingQueueStatusFailed     = "failed"
)

type EmbeddingQueue struct {
	ID           uint       `gorm:"primaryKey" json:"id"`
	TagID        uint       `gorm:"not null;index" json:"tag_id"`
	Status       string     `gorm:"size:20;not null;default:pending;index" json:"status"`
	ErrorMessage string     `gorm:"type:text" json:"error_message"`
	RetryCount   int        `gorm:"default:0" json:"retry_count"`
	CreatedAt    time.Time  `json:"created_at"`
	StartedAt    *time.Time `json:"started_at"`
	CompletedAt  *time.Time `json:"completed_at"`
	Tag          *TopicTag  `gorm:"foreignKey:TagID" json:"tag,omitempty"`
}

func (EmbeddingQueue) TableName() string {
	return "embedding_queues"
}
